### PR TITLE
Range-ify std.path.defaultExtension by adding defaultExt()

### PR DESCRIPTION
**Pull request by WalterBright**
http://dlang.org/phobos/std_path.html#.defaultExtension

100% unit test coverage

### CODE REVIEW

File: std/path.d [hunk 1]
```diff
@@ -1073,19 +1073,11 @@ unittest
     ---
 */
 immutable(Unqual!C1)[] defaultExtension(C1, C2)(in C1[] path, in C2[] ext)
-    @trusted pure // TODO: nothrow (because of to())
+    @trusted
```
Comment (andralex) [inline std/path.d]: This shouldn't be trusted seeing as it's generic, but I guess it's no worse than before. What happened to pure?

File: std/path.d [hunk 2]
```diff
@@ -1073,19 +1073,11 @@ unittest
...
-    if (i == -1)
-    {
-        if (ext.length > 0 && ext[0] == '.')
-            return cast(typeof(return))(path~ext);
-        else
-            return cast(typeof(return))(path~'.'~ext);
-    }
-    else return to!(typeof(return))(path);
+    import std.array;
+    return cast(typeof(return))(defaultExt(path, ext).array);
```
Comment (andralex) [inline std/path.d]: This should solve it:
    return to!(typeof(return))(defaultExt(path, ext));

File: std/path.d [hunk 3]
```diff
@@ -1108,6 +1100,56 @@ unittest
...
+auto defaultExt(R, C)(R path, C[] ext)
+    if ((isRandomAccessRange!R && hasSlicing!R && hasLength!R && isSomeChar!(ElementType!R) ||
+         is(StringTypeOf!R)) &&
+        isSomeChar!C)
+{
+    import std.range : only, chain;
+    import std.utf : byUTF;
+
+    alias CR = Unqual!(ElementEncodingType!R);
+    auto dot = only(cast(CR)'.');
```
Comment (andralex) [inline std/path.d]: use `to` instead of cast

File: std/path.d (hunk 1 above)
Comment (WalterBright) [inline std/path.d]: purity should be inferred

**Comment by andralex** [comment]
Auto-merge toggled on

**Pull request merged**